### PR TITLE
✨ Rework node reuse logic to handle corner cases

### PR DESCRIPTION
**Pull request by furkatgofurov7**
**What this PR does / why we need it**:

This PR addresses handling of some corner cases in node reuse feature, such as:
- avoids cluster deletion problem while testing stack of e2e tests that are run consequitevily by adding a DeletionTimestamp.IsZero() check just before going deep into node reuse logic to avoid fetching non existing resources if any requested while cluster deletion;
- handles m3data re-queuing in cases when machine is deleted during in place upgrade and new machine is immediately failing to get the Metal3Metadata thus breaking the in place upgrade case and also adds needed unit tests.
- improves logging in node reuse

**Note:** These changes were tested using e2e framework tests in #390 and feature tests in [metal3-dev-env/pull/855](https://github.com/metal3-io/metal3-dev-env/pull/855)

### CODE REVIEW

**Comment by mboukhalfa** [comment]
LGTM thanks @furkatgofurov7  for the great work

File: baremetal/metal3machine_manager.go [hunk 1]
```diff
@@ -522,60 +564,66 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
+				}
+				if m3mt != nil {
+					if m3mt.Spec.NodeReuse {
+						if host.Labels == nil {
+							host.Labels = make(map[string]string)
+						}
+						// Check if machine is ControlPlane
+						if m.isControlPlane() {
+							// Fetch KubeadmControlPlane name for controlplane machine
+							m.Log.Info(fmt.Sprintf("Fetch KubeadmControlPlane name while deprovisioning host %v", host.Name))
```
Comment (fmuyassarov) [inline baremetal/metal3machine_manager.go:605]: Just nit: This seems a bit redundant and we have quite many logging in this if already.

File: baremetal/metal3machine_manager.go [hunk 2]
```diff
@@ -522,60 +564,66 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
+							kcpName, err := m.getKubeadmControlPlaneName(ctx)
+							if err != nil {
+								return err
+							}
+							m.Log.Info(fmt.Sprintf("Fetched KubeadmControlPlane name %v", kcpName))
+							// Set nodeReuseLabelName on the host to KubeadmControlPlane name
+							m.Log.Info(fmt.Sprintf("Setting nodeReuseLabelName in host %v to fetched KubeadmControlPlane name %v", host.Name, kcpName))
```
Comment (fmuyassarov) [inline baremetal/metal3machine_manager.go:612]: nit: perhaps these can be merged into one

File: baremetal/metal3machine_manager.go [hunk 3]
```diff
@@ -522,60 +564,66 @@ func (m *MachineManager) Delete(ctx context.Context) error {
...
+						} else {
+							// Fetch MachineDeployment name for worker machine
+							m.Log.Info(fmt.Sprintf("Fetch MachineDeployment name while deprovisioning host %v", host.Name))
+							mdName, err := m.getMachineDeploymentName(ctx)
+							if err != nil {
+								return err
+							}
+							m.Log.Info(fmt.Sprintf("Fetched MachineDeployment name %v", mdName))
+							// Set nodeReuseLabelName on the host to MachineDeployment name
+							m.Log.Info(fmt.Sprintf("Setting nodeReuseLabelName in host %v to fetched MachineDeployment name %v", host.Name, mdName))
```
Comment (fmuyassarov) [inline baremetal/metal3machine_manager.go:623]: same here with regard to merging

**Review by fmuyassarov — COMMENTED** [review]
looks good, just some nit comments left inline

File: baremetal/metal3machine_manager.go (hunk 1 above)
Comment (furkatgofurov7) [reply baremetal/metal3machine_manager.go:605]: I think this is helpful to know which BMH being deprovisioned and in which BMH we expect nodeReuse label while debugging

File: baremetal/metal3machine_manager.go [hunk 4]
```diff
@@ -390,6 +392,46 @@ func (m *MachineManager) Associate(ctx context.Context) error {
...
+			m.SetError("Failed to get the BaremetalHost for the Metal3Machine",
+				capierrors.CreateMachineError,
+			)
+			return err
+		}
+
+		if err = m.setHostSpec(ctx, host); err != nil {
+			if _, ok := err.(HasRequeueAfterError); !ok {
+				m.SetError("Failed to set the BaremetalHost Specs",
+					capierrors.CreateMachineError,
```
[the remark this answers is not in the document]
Comment (furkatgofurov7) [reply baremetal/metal3machine_manager.go:415]: Yeah, agree, but at the same time if you check the manager code, you would find good number of `capierrors.CreateMachineError` calls in different use cases where we fail to get/set/associate/annotate certain objects, so that was the reason I followed the same pattern.

**Review by namnx228 — COMMENTED** [review]
LGTM

**Comment by furkatgofurov7** [comment]
cc @Xenwar
